File: api/app/services/recurring_schedule.py
import calendar
from datetime import datetime, timedelta, timezone


def utcnow() -> datetime:
    return datetime.now(timezone.utc)


def ensure_aware(dt: datetime) -> datetime:
    if dt.tzinfo is None:
        return dt.replace(tzinfo=timezone.utc)
    return dt


def add_months(dt: datetime, months: int) -> datetime:
    month = dt.month - 1 + months
    year = dt.year + month // 12
    month = month % 12 + 1
    day = min(dt.day, calendar.monthrange(year, month)[1])
    return dt.replace(year=year, month=month, day=day)


def advance_next_run(current: datetime, frequency: str) -> datetime:
    current = ensure_aware(current)
    if frequency == "daily":
        return current + timedelta(days=1)
    if frequency == "weekly":
        return current + timedelta(weeks=1)
    if frequency == "monthly":
        return add_months(current, 1)
    if frequency == "yearly":
        return add_months(current, 12)
    raise ValueError(f"Unknown frequency: {frequency}")


def compute_initial_next_run(anchor: datetime, frequency: str, *, now: datetime | None = None) -> datetime:
    candidate = ensure_aware(anchor)
    reference = ensure_aware(now or utcnow())
    while candidate <= reference:
        candidate = advance_next_run(candidate, frequency)
    return candidate

File: apps/mobile/src/services/recurringSchedule.ts
export type RecurringFrequency = "daily" | "weekly" | "monthly" | "yearly";

export function addMonths(date: Date, months: number): Date {
  const d = new Date(date);
  const day = d.getDate();
  d.setDate(1);
  d.setMonth(d.getMonth() + months);
  const lastDay = new Date(d.getFullYear(), d.getMonth() + 1, 0).getDate();
  d.setDate(Math.min(day, lastDay));
  return d;
}

export function advanceNextRun(current: Date, frequency: RecurringFrequency): Date {
  const next = new Date(current);
  if (frequency === "daily") {
    next.setDate(next.getDate() + 1);
    return next;
  }
  if (frequency === "weekly") {
    next.setDate(next.getDate() + 7);
    return next;
  }
  if (frequency === "monthly") {
    return addMonths(next, 1);
  }
  return addMonths(next, 12);
}

export function computeInitialNextRun(anchor: Date, frequency: RecurringFrequency, now = new Date()): Date {
  let candidate = new Date(anchor);
  candidate.setHours(9, 0, 0, 0);
  while (candidate <= now) {
    candidate = advanceNextRun(candidate, frequency);
  }
  return candidate;
}

export function anchorFromDateIso(dateIso: string): Date {
  const [y, m, d] = dateIso.split("-").map(Number);
  const anchor = new Date(y, m - 1, d, 9, 0, 0, 0);
  return anchor;
}

export function frequencyLabelKey(frequency: RecurringFrequency): string {
  return `recurring.freq.${frequency}`;
}
